src/courses/mappers/09.md

```diff
@@ -0,0 +1,108 @@
+---
+order: 9
+next: 10.md
+title: Mapper Preparations - OHDF Converters
+author: Charles Hu
+---
+
+## An Overview of OHDF Converters
+
+OHDF Converters is the library that hosts and provides OHDF mappers to other applications in the SAF tool suite. It is important to have a general understanding of a few key features of OHDF Converters as the mappers you develop will be integrated here and leverage the library's utilities.
+
+### Features
+
+- Implements OHDF mappers to transform security data formats to and from OHDF.
+- Special keyword functions provided by the <i>base-converter</i> class.
+- Fingerprinting services for automatic data format identification.
+
+### Keywords
+
+<i>base-converter</i> is an underlying class which enables \*-to-HDF mapping in OHDF Converters. It defines \*-to-HDF mappers and provides critical tools in the form of keywords which allow for the construction of such mappers. All \*-to-HDF mappers inherit from this class and therefore have access to the keywords that this class provides. You should aim to utilize these keywords to their fullest potential to ease and simplify mapper development. The provided tools are as follows:
+
+- `path`: Define JSON object path to go to. Paths are found recursively.
+  - Use:
+  ```typescript
+  path: PATH_AS_STRING
+  ```
+  - Example:
+  ```typescript
+  // Attribute 'id' will be set as whatever JSON object attribute 'vulnerability.id' is
+  id: {path: 'vulnerability.id'}
+  ```
+
+- `transformer`: Execute given code sequence; operates similar to an anonymous function.
+  - Use:
+  ```typescript
+  transformer: (PARAMETER: TYPE): OUTPUT_TYPE => {CODE_TO_EXECUTE}
+  ```
+  - Example:
+  ```typescript
+  // Attribute 'cci' will be set as the returned CCI tag from the ingested 'data' argument
+  cci: {
+    path: 'vulnerabilityClassifications',
+    transformer: (data: string) => getCCIsForNISTTags(nistTag(data))
+  }
+  ```
+
+- `arrayTransformer`: Execute given code sequence on a given array; primarily used when in an attribute that is an array of objects.
+  - Use:
+  ```typescript
+  arrayTransformer: (PARAMETER: TYPE): OUTPUT_TYPE => {CODE_TO_EXECUTE}
+  ```
+  - Example:
+  ```typescript
+  // The function 'deduplicateId' will run against all items in the current array that the 'arrayTransformer' was called inside
+  arrayTransformer: deduplicateId
+  ```
+
+- `pathTransform`: Converts the JSON object path structure.
+  - Use:
+  ```typescript
+  pathTransform: (PARAMETER: TYPE): OUTPUT_TYPE => {CODE_TO_EXECUTE}
+  ```
+  - Example:
+  ```typescript
+  // Returns the JSON path if it is an array, otherwise returns an empty array
+  pathTransform: (value) => (Array.isArray(value) ? value : []),
+  ```
+
+- `key`: Used by <i>base-converter</i> to sort an array of objects by
```

Comment:
This doesn't do any sorting or anything in and of itself.  This keyword currently can only be used in the controls array area.  Its purpose is to automatically handle doing the dedup for the controls array to ensure that they are unique, so its value currently should only ever be 'id'.  When calling 'super' in the constructor for the mapper (to call the base-converter constructor, which sets up certain variables we need), it takes in a boolean parameter called "collapseResults".  This parameter is used during the auto-dedup function - on whether the dedup process for the control array should try to dedup the results array when it is deduping the controls.